components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * <p>Takes care of:
+ * <ol>
+ *   <li>setting up Unix Domain Socket files for each pluggable component,</li>
+ *   <li>environment variable handling,</li>
+ *   <li>setting up the required server machinery to expose the service and handle requests for it.</li>
+ * </ol>
+ *
+ * <p>If you end up needing to extend functionality of this class it might be simpler re-implementing it.
+ */
+public class PluggableComponentServer {
+  private static final Logger log = Logger.getLogger(PluggableComponentServer.class.getName());
+
+  /**
+   * The folder where UDS files will be created.
+   */
+  private final String unixDomainSocketFolder;
+
+  // Our "list" or registered components. We use a Map to avoid duplicated registrations.
+  private final Map<String, Server> servers = new HashMap<>();
+
+  /**
+   * Default Constructor.
+   */
+  public PluggableComponentServer() {
+    this.unixDomainSocketFolder = getUnixSocketFolderOrAbort();
+  }
+
+  /**
+   * Add a new pluggable component to this server.
+   *
+   * @param component the pluggable component to be added.
+   * @return A reference to this {@link PluggableComponentServer}.
+   * @throws IOException if we face problems creating the Unix Domain Socket file for this component.
+   */
+  public PluggableComponentServer registerComponent(final PluggableComponent component) throws IOException {
+    final String componentName = component.getName();
+    if (servers.containsKey(componentName)) {
+      logFatalAndAbort("A pluggable component named '" + componentName
+          + "' was already registered. Aborting.");
+    }
+    final Server componentServer = buildServerForComponent(component);
+
+    this.servers.put(componentName, componentServer);
+
+    return this;
+  }
+
+  /**
+   * Starts all pluggable component servers and wait for them to complete.
+   *
+   * @throws IOException if we face errors starting the servers for each component.
+   */
+  public void run() throws IOException {
+    startServers();
+    blockUntilShutdown();
+  }
+
+  //
+  // Component Registration
+  //
+
+  private Server buildServerForComponent(final PluggableComponent component) throws IOException {
+    final String componentName = component.getName();
+    log.info("Creating server for component " + componentName);
+
+    // Sanity check: we can't do anything with a pluggable component if it isn't exposing any service.
+    final List<BindableService> exposedServices = component.getExposedServices();
+    if (exposedServices.isEmpty()) {
+      logFatalAndAbort("Pluggable component " + componentName
+          + " is not exposing any service or Dapr API. Aborting.");
+    }
+
+    // Unix Domain Socket setup
+    final var componentUdsPath = buildPathForComponentUnixDomainSocket(componentName);
+    log.info("Configuring server to listen to unix socket domain on file " + componentUdsPath.toAbsolutePath());
+    // If file exists, remove it.
+    final File componentUdsFile = componentUdsPath.toFile();
```

Comment:
Does Java need to update the permissions of the UDS file post-creation like in .NET due to permissions differences between the sidecar and component?

Reply:
Yes, it it need - umask on macos by default is pretty lax! Got catch!

Reply:
This is now addressed on `fixUnixDomainSocketFilePermissions()`, called after the `server.start()`, as the UDS is only created then: here we are just doing sanity checks and cleanups.